File: clients/apple/MultiLiveTV/Services/SourceStore.swift
import Foundation

final class SourceStore {
    private let sources: [Source]
    private let byNumericId: [Int: Source]
    private let bySourceIdKey: [String: Source]

    init(bundle: Bundle = .main) throws {
        let document = try Self.loadRegistry(bundle: bundle)
        sources = document.sources
        byNumericId = Self.uniqueIndex(sources, key: \.numericId)
        bySourceIdKey = Self.uniqueIndex(sources, key: \.sourceId)
    }

    init(sources: [Source]) {
        self.sources = sources
        byNumericId = Self.uniqueIndex(sources, key: \.numericId)
        bySourceIdKey = Self.uniqueIndex(sources, key: \.sourceId)
    }

    private static func uniqueIndex<Key: Hashable>(
        _ sources: [Source],
        key: KeyPath<Source, Key>
    ) -> [Key: Source] {
        Dictionary(sources.map { ($0[keyPath: key], $0) }, uniquingKeysWith: { _, last in last })
    }

    func all() -> [Source] { sources }

    /// Configured entry regardless of enabled (deep link / diagnostics).
    func configured(id numericId: Int) -> Source? {
        byNumericId[numericId]
    }

    func configured(sourceId: String) -> Source? {
        bySourceIdKey[sourceId]
    }

    /// Admin-enabled and not vip-gated — may still be unhealthy.
    func enabled() -> [Source] {
        sources.filter { $0.enabled && ($0.vipOnly != true) }
    }

    /// Enabled sources that may be used for a given capability.
    func collectable(capability: KeyPath<SourceCapabilities, Bool>) -> [Source] {
        enabled().filter { $0.capabilities[keyPath: capability] }
    }

    /// Collectable by numeric id; nil when disabled / missing / vip.
    func byID(_ id: Int) -> Source? {
        guard let source = byNumericId[id], source.enabled, source.vipOnly != true else {
            return nil
        }
        return source
    }

    func bySourceID(_ sourceId: String) -> Source? {
        guard let source = bySourceIdKey[sourceId], source.enabled, source.vipOnly != true else {
            return nil
        }
        return source
    }

    func `default`() -> Source? {
        enabled().first
    }

    func metadataPriority(for numericId: Int) -> Int {
        configured(id: numericId)?.priority.metadataPriority ?? 100
    }

    func playPriority(for numericId: Int) -> Int {
        configured(id: numericId)?.priority.playPriority ?? 100
    }

    private static func loadRegistry(bundle: Bundle) throws -> SourceRegistryDocument {
        if let url = bundle.url(forResource: "source-registry", withExtension: "json") {
            let data = try Data(contentsOf: url)
            return try JSONDecoder().decode(SourceRegistryDocument.self, from: data)
        }
        // Legacy fallback for older bundles / tests without registry resource.
        guard let url = bundle.url(forResource: "sources", withExtension: "json") else {
            throw VodError.missingSources
        }
        let data = try Data(contentsOf: url)
        let legacy = try JSONDecoder().decode([LegacySource].self, from: data)
        return SourceRegistryDocument(
            version: 0,
            sources: legacy.map { $0.toSource() }
        )
    }
}

private struct LegacySource: Codable {
    let id: Int
    let name: String
    let url: String
    let flag: Int
    let jxUrl: String?
    let vipOnly: Bool?

    enum CodingKeys: String, CodingKey {
        case id, name, url, flag
        case jxUrl = "jx_url"
        case vipOnly = "vip_only"
    }

    func toSource() -> Source {
        Source(
            id: id,
            name: name,
            url: url,
            flag: flag,
            jxUrl: jxUrl,
            vipOnly: vipOnly
        )
    }
}

enum VodError: LocalizedError {
    case missingSources
    case sourceNotFound
    case vodNotFound

    var errorDescription: String? {
        switch self {
        case .missingSources: return "未找到 source-registry.json"
        case .sourceNotFound: return "资源站不存在"
        case .vodNotFound: return "未找到影片"
        }
    }
}

File: clients/android/core/src/main/kotlin/com/heibaimiao/multilivetv/source/SourceStore.kt
package com.heibaimiao.multilivetv.source

import com.heibaimiao.multilivetv.model.LegacySource
import com.heibaimiao.multilivetv.model.Source
import com.heibaimiao.multilivetv.model.SourceCapabilities
import com.heibaimiao.multilivetv.model.SourceRegistryDocument
import com.heibaimiao.multilivetv.model.VodError
import com.heibaimiao.multilivetv.net.VodClientException
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

enum class SourceCapability {
    SEARCH, CATEGORY, DETAIL, PLAY, PAGINATION, LIVE
}

class SourceStore(private val sources: List<Source>) {
    private val byNumericId = sources.associateBy { it.numericId }
    private val bySourceIdKey = sources.associateBy { it.sourceId }

    constructor() : this(loadRegistry().sources)

    fun all(): List<Source> = sources

    fun configured(id: Int): Source? = byNumericId[id]

    fun configured(sourceId: String): Source? = bySourceIdKey[sourceId]

    fun enabled(): List<Source> = sources.filter { it.enabled && it.vipOnly != true }

    fun collectable(capability: SourceCapability): List<Source> = enabled().filter { source ->
        when (capability) {
            SourceCapability.SEARCH -> source.capabilities.search
            SourceCapability.CATEGORY -> source.capabilities.category
            SourceCapability.DETAIL -> source.capabilities.detail
            SourceCapability.PLAY -> source.capabilities.play
            SourceCapability.PAGINATION -> source.capabilities.pagination
            SourceCapability.LIVE -> source.capabilities.live
        }
    }

    fun byID(id: Int): Source? {
        val source = byNumericId[id] ?: return null
        if (!source.enabled || source.vipOnly == true) return null
        return source
    }

    fun bySourceID(sourceId: String): Source? {
        val source = bySourceIdKey[sourceId] ?: return null
        if (!source.enabled || source.vipOnly == true) return null
        return source
    }

    fun default(): Source? = enabled().firstOrNull()

    fun metadataPriority(numericId: Int): Int =
        configured(numericId)?.priority?.metadataPriority ?: 100

    fun playPriority(numericId: Int): Int =
        configured(numericId)?.priority?.playPriority ?: 100

    companion object {
        private val json = Json { ignoreUnknownKeys = true; isLenient = true }

        fun loadRegistry(
            registryText: String? = resourceOrNull("source-registry.json"),
            legacyText: String? = resourceOrNull("sources.json"),
        ): SourceRegistryDocument {
            if (registryText != null) {
                return json.decodeFromString(SourceRegistryDocument.serializer(), registryText)
            }
            if (legacyText != null) {
                val legacy = json.decodeFromString(ListSerializer(LegacySource.serializer()), legacyText)
                return SourceRegistryDocument(version = 0, sources = legacy.map { it.toSource() })
            }
            throw VodClientException(VodError.MISSING_SOURCES.message)
        }

        private fun resourceOrNull(name: String): String? {
            val stream = SourceStore::class.java.classLoader?.getResourceAsStream(name) ?: return null
            return stream.bufferedReader(Charsets.UTF_8).use { it.readText() }
        }
    }
}

fun Source.supports(capability: SourceCapability): Boolean = when (capability) {
    SourceCapability.SEARCH -> capabilities.search
    SourceCapability.CATEGORY -> capabilities.category
    SourceCapability.DETAIL -> capabilities.detail
    SourceCapability.PLAY -> capabilities.play
    SourceCapability.PAGINATION -> capabilities.pagination
    SourceCapability.LIVE -> capabilities.live
}

fun SourceCapabilities.has(capability: SourceCapability): Boolean = when (capability) {
    SourceCapability.SEARCH -> search
    SourceCapability.CATEGORY -> category
    SourceCapability.DETAIL -> detail
    SourceCapability.PLAY -> play
    SourceCapability.PAGINATION -> pagination
    SourceCapability.LIVE -> live
}
